src/D2L.CodeStyle.Analyzers/Immutability/MutabilityInspector.cs

```diff
@@ -549,6 +555,75 @@ HashSet<ITypeSymbol> typeStack
 			}
 		}
 
+		private MutabilityInspectionResult PerformGenericArgumentInspection(
+			ITypeSymbol type,
+			HashSet<ITypeSymbol> typeStack
+		) {
+			MutabilityInspectionResult classArgumentResult =
+				InspectGenericArguments( type, typeStack );
+			if( classArgumentResult.IsMutable ) {
+				return classArgumentResult;
+			}
+
+			MutabilityInspectionResult interfaceArgumentResult =
+				InspectInterfaceArguments( type, typeStack );
+			if( interfaceArgumentResult.IsMutable ) {
+				return interfaceArgumentResult;
+			}
+
+			return MutabilityInspectionResult.NotMutable();
+		}
+
+		private MutabilityInspectionResult InspectInterfaceArguments(
+			ITypeSymbol type,
+			HashSet<ITypeSymbol> typeStack
+		) {
+
+			foreach( INamedTypeSymbol intf in type.Interfaces ) {
+				if( intf.IsGenericType() ) {
+					var result = InspectGenericArguments( intf, typeStack );
+					if( result.IsMutable ) {
+						return result;
+					}
+				}
+			}
+
+			return MutabilityInspectionResult.NotMutable();
+		}
+
+		private MutabilityInspectionResult InspectGenericArguments(
+			ITypeSymbol type,
+			HashSet<ITypeSymbol> typeStack
+		) {
+			if( type.IsGenericType() ) {
+				var symbolType = type as INamedTypeSymbol;
+				if( symbolType == default ) {
+					return MutabilityInspectionResult.MutableType( type, MutabilityCause.IsPotentiallyMutable );
+				}
+
+				int index = 0;
+				foreach( ITypeSymbol argument in symbolType.TypeArguments ) {
+					ImmutabilityScope argumentScope =
+						symbolType.TypeParameters[index].GetImmutabilityScope();
+
+					if( argumentScope == ImmutabilityScope.SelfAndChildren ) {
+						MutabilityInspectionResult result = DoInspectType(

```

Comment:
I would have moved these all out in to the ImmutableGenericArgumentClass, but this needs to be able to call DoInspectType and have the type stack available.